### DISCUSSION TITLE
(NuGet) How to make a cross-platform library for MonoGame?

### DISCUSSION THREAD

**Discussion by ymiroshnyk** [MonoGame/MonoGame#8422] [General]
Hi. I’m developing the [NeutrinoParticles](https://neutrinoparticles.com/) project that makes particle effects on different platforms including MonoGame.

For MonoGame I have a renderer library, that basically was unified for OpenGL and DirectX. This library is published at NuGet as NeutrinoParticles.MonoGame. And before it has a dependency on MonoGame.Framework.Portable, which allowed it to be a unified library for DirectX and OpenGL.

During the latest upgrade to .NET 6.0 I found the Portable package is obsolete. So my question is, how do I correctly make a cross-platform library for MonoGame (at least DX & GL)?

In Visual Studio, I found a template for MonoGame Game Library which is considered to be a cross-platform library, but it has a dependency on MonoGame.Framework.DesktopGL that doesn’t work with DirectX.

**Comment by ThomasFOG** [comment]
It depends on what you mean with "cross-platform" because it seems that you are confusing platforms and rendering backends.

If you want your package to build and run on all systems, you can simply reference DesktopGL. It'll work anywhere (unlike WindowsDX which will run only on Windows systems).

If you want your package to be able to switch rendering backends, you can't with the current architecture and you'll have to make a package for each backends.

Portable stuff have indeed been deprecated and the current cross-platform route is to just use DesktopGL.

**Reply by ymiroshnyk** [reply]
Not really switch backends, but to let the main application that uses the library to define the backend and the library will only detect it and use correct effects for render. The rest of the code in the library is the same for any rendering backend.

If it is not possible, I believe I can use shared library to share most of the code between NeutrinoParticles.MonoGame.DX and NeutrinoParticles.MonoGame.GL.

**Reply by ThomasFOG** [reply]
You can't do that at runtime.

**Comment by Apostolique** [comment]
You can check how I set up my libraries. A simple one is <https://github.com/Apostolique/Apos.Camera>.

The main thing to look into is the .csproj in the Source folder.

If you want to try my setup, I made some templates that you can use:

```
dotnet new install Apos.Framework.Library.CSharp
dotnet new aposframeworklibrary -o Apos.Camera --repo https://github.com/Apostolique/Apos.Camera --branch main --description "Camera library for MonoGame." --param:author "Jean-David Moisan"
```

The files are here: <https://github.com/Apostolique/Apos.Framework/tree/main/Source/Library>.

Here is a more complex one that can build a shader for both DesktopGL and WindowsDX: <https://github.com/Apostolique/Apos.Shapes>.
